src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,173 @@
+import * as React from "react";
+
+import { OBSCURE_KEY, OBSCURE_PASSCODE, PASSWORD_KEY } from "../constants";
+import createPersistedState from "../hooks/usePersistedState";
+
+type AuthState = "unauthenticated" | "authenticated" | "obscured";
+
+type AuthSetters =
+  | { type: "passcode"; value: string | null }
+  //This is to set up the future use of the obscure pass beings set by the user. if the `value` is undefined, the default obscure pass is being used
+  | { type: "obscure"; value?: string | null };
+
+type AuthValuesSet = {
+  passcodeSet: boolean;
+  obscureSet: boolean;
+};
+
+type SecurityContextType = {
+  authValuesSet: AuthValuesSet;
+  setAuthValues: (val: AuthSetters) => void;
+  authenticate: (val: string | null, validateOnly?: boolean) => boolean;
+  authState: AuthState;
+};
+
+const DefaultState: SecurityContextType = {
+  authValuesSet: { passcodeSet: false, obscureSet: false },
+  setAuthValues: () => {},
+  authenticate: () => false,
+  authState: "unauthenticated",
+};
+
+export const SecurityContext = React.createContext<SecurityContextType>(
+  DefaultState
+);
+
+const usePersistedPasscodeState = createPersistedState(PASSWORD_KEY);
+const usePersistedObscureState = createPersistedState(OBSCURE_KEY);
+
+export const SecurityProvider = ({
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
+  const [passcode, status, setPasscode] = usePersistedPasscodeState<
+    string | null
+  >(null);
+  if (status === "loading") return null;
+  return (
+    <SecurityProviderInner passcode={passcode} setPasscode={setPasscode}>
+      {children}
+    </SecurityProviderInner>
+  );
+};
+
+const SecurityProviderInner = ({
+  children,
+  passcode,
+  setPasscode,
+}: {
+  children: React.ReactNode;
+  passcode: string | null;
+  setPasscode: React.Dispatch<React.SetStateAction<null | string>>;
+}) => {
+  const [authState, setAuthState] = React.useState<AuthState>(
+    passcode === null ? "authenticated" : "unauthenticated"
+  );
+
+  const [obscureCode, , setObscureCode] = usePersistedObscureState<
+    string | null
+  >(null);
+
+  const setPasscodeWithValidation = React.useCallback(
+    (passcodeValue: string | null) => {
+      // If user is able to set their own obscure passcode get rid of this if statement.
+      if (passcodeValue === OBSCURE_PASSCODE) {
+        throw new Error("passcode is reserved");
+      }
+
+      if (passcode !== null && passcodeValue === obscureCode) {
+        throw new Error("passcode is already being used obscure code");
+      }
+
+      if (!validPasscode(passcodeValue)) {
```

Comment:
This is a non-blocking comment:

A pattern that can work well for this is to change this to a `validatePasscode` function that throws an error, rather than something that returns a boolean. That way you can do other checks like the one above (`passcodeValue === obscureCode`), and you can throw a more meaningful error message than just "passcode not valid", e.g. it can be "passcode must be at least 5 characters". For all these errors though you will probably need to define an error code, so that the text can be translated in the frontend. At some point we will need some code to put all the errors in one place, but that is not a priority right now I think.